README.md
# Rich Theme Manager

## Description

Implements a basic "theme manager" class for managing rich [Themes](https://rich.readthedocs.io/en/stable/style.html#style-themes) in your [rich](https://github.com/Textualize/rich) CLI application.

The rich package provides an easy way to define custom styles and themes for use with rich.  This package provides a simple way to manage the themes: e.g. to list, add, remove themes, for the user to preview themes, and to manage themes on disk.

## What problem does this solve?

Consider this scenario from the rich [documentation](https://rich.readthedocs.io/en/stable/style.html#style-themes):

If you re-use styles it can be a maintenance headache if you ever want to modify an attribute or color – you would have to change every line where the style is used. Rich provides a Theme class which you can use to define custom styles that you can refer to by name. That way you only need to update your styles in one place.

Style themes can make your code more semantic, for instance a style called "warning" better expresses intent that "italic magenta underline".

To use a style theme, construct a Theme instance and pass it to the Console constructor. Here’s an example:

```python
from rich.console import Console
from rich.theme import Theme
custom_theme = Theme({
    "info": "dim cyan",
    "warning": "magenta",
    "danger": "bold red"
})
console = Console(theme=custom_theme)
console.print("This is information", style="info")
console.print("[warning]The pod bay doors are locked[/warning]")
console.print("Something terrible happened!", style="danger")
```

I highly recommend the use of Themes in your rich application instead of hard-coding colors and styles. However, there's still a problem of managing themes.  For example, letting the user of your application change the default theme styles you've chosen.  What if they are color blind and can't distinguish the colors you've selected? What if they're using a light terminal background and you've selected colors best suited for a dark terminal?  

The rich `Theme` class provides everything you to manage these situations but doing so requires a fair amount of boiler plate code for each application. This package attempts to provide an easy solution for these scenarios with a minimal amount of code.  Instead of having to implement management of theme files and allowing the user to list or preview themes yourself, you can use the `ThemeManager` class to manage themes for you.

## Synopsis

Using `rich_theme_manager` is easy and takes just a few lines of code.  Import `Theme` from `rich_theme_manager` instead of from `rich.theme` (`rich_theme_manager.Theme` subclasses `rich.theme.Theme`) then use `ThemeManager` to manage themes.  `ThemeManager` can be created with or without a `theme_dir` argument. If you don't provide `theme_dir`, `ThemeManager` will not manage themes on disk.  This may still be useful for allowing the user to list and preview themes.  If you do provide `theme_dir`, any default themes passed to `ThemeManager` will be written to `theme_dir` and any theme config files found in `theme_dir` will be loaded.

```python
from rich.console import Console
from rich.style import Style
import pathlib

from rich_theme_manager import Theme, ThemeManager


THEMES = [
    Theme(
        name="dark",
        description="Dark mode theme",
        tags=["dark"],
        styles={
            "info": "dim cyan",
            "warning": "bold magenta",
            "danger": "bold red",
        },
    ),
    Theme(
        name="light",
        description="Light mode theme",
        styles={
            "info": Style(color="#22863a", bold=True),
            "warning": Style(color="#032f62", bold=True),
            "danger": Style(color="#b31d28", bold=True, underline=True, italic=True),
        },
    ),
    Theme(
        name="mono",
        description="Monochromatic theme",
        tags=["mono", "colorblind"],
        styles={
            "info": "italic",
            "warning": "bold",
            "danger": "reverse bold",
        },
    ),
]

if __name__ == "__main__":
    # you can specify a config directory to save/load themes to/from
    theme_dir = pathlib.Path("~/.rich_theme_manager/themes").expanduser()
    theme_dir.expanduser().mkdir(parents=True, exist_ok=True)

    theme_manager = ThemeManager(theme_dir=theme_dir, themes=THEMES)
    theme_manager.list_themes()
    print("\n")

    dark = theme_manager.get("dark")
    theme_manager.preview_theme(dark)
    console = Console(theme=dark)
    print("\n")

    console.print("This is information", style="info")
    console.print("[warning]The pod bay doors are locked[/warning]")
    console.print("Something terrible happened!", style="danger")
```

![Example output](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/example1.png)

## Example app

A simple example app that demonstrates the ThemeManager class comes with rich_theme_manager in `__main__.py`:

`python -m rich_theme_manager`:

```text
usage: rich_theme_manager [-h] [--example [EXAMPLE]] [--list] [--preview THEME] [--config THEME]

Example CLI usage of rich_theme_manager

optional arguments:
  -h, --help           show this help message and exit
  --example [EXAMPLE]  Show example output for theme.
  --list               List themes.
  --preview THEME      Preview theme.
  --config THEME       Print configuration for theme THEME.
```

`python -m rich_theme_manager --list`:

![Example --list output](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/list.png)

`python -m rich_theme_manager --preview dark`:

![Example --preview output](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/preview_dark.png)

## Documentation

### Theme class

`rich_theme_manager.Theme` is a subclass of `rich.theme.Theme` and provides additional functionality for managing themes.

```python
    Theme(
          name: str,
          description: Optional[str] = None,
          styles: Optional[Mapping[str, StyleType]] = None,
          inherit: bool = True,
          tags: Optional[List[str]] = None,
          path: Optional[str] = None,
    )
```

Arguments

* `name`: The name of the theme; required
* `description`: An optional description of the theme.
* `styles`: An optional mapping of style names to styles.
* `inherit`: Whether the theme inherits from the default theme.
* `tags`: An optional list of tags for the theme; useful for allowing user to filter themes.
* `path`: The path to the theme file; in normal use this is not needed as `ThemeManager` will automatically create the theme file.

Properties

* `Theme().name`: The name of the theme
* `Theme().description`: Description of the theme
* `Theme().tags`: List of tags for the theme
* `Theme().inherit`: bool indicating whether the theme inherits from the default theme
* `Theme().style_names`: List of names for styles in the theme
* `Theme().path`: The path to the theme file; (getter/setter)
* `Theme().config`: Contents of a configuration file for the theme (same format as `rich.theme.Theme().config` but with an additional `[metadata]` section)

Methods

* `Theme().save()`: Save the theme to disk (to `Theme().path`)
* `Theme().load()`: Load the theme from disk (from `Theme().path`)
* `Theme().to_file(path: str)`: Save the theme to disk (to `path`)
* `Theme().update(other: Theme, overwrite_existing_styles: bool = True)`: Update the theme with the values from another theme, optionally overwriting existing styles.

Class Methods:

* `Theme.from_file(config_file: IO[str], source: Optional[str] = None, inherit: bool = True)` -> `Theme`: Load a theme from a text mode configuration file (in [configparser](https://docs.python.org/3/library/configparser.html) INI format).
* `Theme.read(path: str, inherit: bool = True) -> Theme`: Load a theme from disk (from `path`)

The `.theme` INI file format looks like this:

```ini
[metadata]
name = dark
description = Dark mode theme
tags = dark
inherit = True

[styles]
danger = bold red
info = dim cyan
warning = bold magenta
```

Here's an real world example of a theme INI file from one of my [apps](https://github.com/RhetTbull/osxphotos):

```INI
[metadata]
name = dark
description = Dark mode theme
tags = dark
inherit = True

[styles]
bar.back = rgb(68,71,90)
bar.complete = rgb(249,38,114)
bar.finished = rgb(80,250,123)
bar.pulse = rgb(98,114,164)
color = rgb(248,248,242)
count = rgb(139,233,253)
error = bold rgb(255,85,85)
filename = bold rgb(189,147,249)
filepath = bold rgb(80,250,123)
highlight = bold #000000 on #d73a49
num = bold rgb(139,233,253)
progress.elapsed = rgb(139,233,253)
progress.percentage = rgb(255,121,198)
progress.remaining = rgb(139,233,253)
time = bold rgb(139,233,253)
uuid = rgb(255,184,108)
warning = bold rgb(241,250,140)
```

`Theme` implements the [rich Console protocol](https://rich.readthedocs.io/en/stable/protocol.html) which means that you use `rich.print()` and `rich.console.Console().print()` to print a theme to the console.  Doing so results in a preview of the theme which visually shows the colors and styles used in the theme.

![Theme preview](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/theme_preview_print.png)

The `Theme` preview will use default sample text for each style. You can change the sample text by setting the `rich_theme_manager.theme.SAMPLE_TEXT` global variable.

![Theme preview with sample text](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/theme_preview_sample_text.png)

`Theme` implements the `__eq__` method so two `Theme` instances can be easily compared for equality. `Theme` instances are considered equal if all properties with exception of `path` are equal.

### ThemeManager class

```python
    ThemeManager(
        theme_dir: Optional[str] = None,
        themes: Optional[List[Theme]] = None,
        overwrite: bool = False,
        update: bool = False,
    )
```

Arguments:

* `theme_dir`: Optional directory to save/load themes to/from.
* `themes`: Optional list of Theme objects
* `overwrite`: overwrite existing theme files
* `update`: update existing theme files with *new* styles from themes but don't replace *existing* styles

If provided, `theme_dir` must exist.  If `theme_dir` is set (for example, using [click.get_app_dir](https://click.palletsprojects.com/en/8.0.x/api/?highlight=get_app_dir#click.get_app_dir)), upon initialization `ThemeManager` will save any default themes provided via `themes` to `theme_dir` and load any themes from `theme_dir`.  Theme files are standard INI files as created by [configparser](https://docs.python.org/3/library/configparser.html) and are named `<name>.theme` where `<name>` is the name of the Theme (see `Theme.name`).  If a theme file already exists, it will be loaded and `ThemeManager` will not overwrite it unless `overwrite=True`.  If `update=True`, any new styles defined in `themes` will be added to the theme file but existing styles will not be replaced. If tags and description of themes are different and `update=True`, those items will also be updated in the theme file.  This allows you to add new styles to an existing theme in future app updates without overriding changes made by the user to theme styles.

Properties:

* `ThemeManager().themes`: List of themes

Methods:

* `ThemeManager().add(theme: Theme, overwrite=False) -> None`: Add a theme to the list of managed themes.  If `overwrite` is True, the theme file will be overwritten if it already exists.
* `ThemeManager().remove(theme: Theme) -> None`: Remove a theme from the list of managed themes and delete the theme file if it exists.
* `ThemeManager().get(theme_name: str) -> Theme`: Get a theme by name. Raises 'ValueError` if no theme with the given name is found.
* `ThemeManager().load_themes(theme_dir=None) -> None`: Load themes from `theme_dir` (or `ThemeManager().theme_dir` if not provided).  Any `.theme` files found in `theme_dir` will be loaded and added to the list of managed themes.
* `ThemeManager().write_themes(overwrite=False) -> None`: Write themes to file (as specified in each `Theme().path` which will be set automatically by `ThemeManager`).  If `overwrite` is True, the theme file will be overwritten if it already exists.

* `ThemeManager().list_themes(show_path: bool = True, theme_names: Optional[List[str]] = None, console: Optional[Console] = None) -> None`: Print a list of themes to the console.  If `show_path` is True, the path to the theme file will be printed.  If `theme_names` is provided, only themes with names in the list will be printed. An optional `rich.console.Console()` instance may be provided to print to a specific console.

![ThemeManager list example](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/theme_manager_list.png)

Class Methods:

* `ThemeManager.preview_theme(theme: Theme, sample_text: Optional[str] = None, show_path: bool = True, console: Optional[Console] = None) -> None`: Print a preview of the theme to the console showing the style of each style in the theme.  If `sample_text` is provided, it will be used as the sample text to preview otherwise a default string will be used.  If `show_path` is True, the path to the theme file will be printed.  An optional `rich.console.Console()` instance may be provided to print to a specific console.

![ThemeManager preview example](https://github.com/RhetTbull/rich_theme_manager/raw/main/images/theme_manager_preview.png)

## Test Coverage

100% coverage of all code with exception of the example CLI app.

```text
---------- coverage: platform darwin, python 3.8.12-final-0 ----------
Name                             Stmts   Miss  Cover
----------------------------------------------------
rich_theme_manager/__init__.py       5      0   100%
rich_theme_manager/manager.py       73      0   100%
rich_theme_manager/theme.py        155      0   100%
----------------------------------------------------
TOTAL                              233      0   100%
```

## License

MIT License

## Contributing

Contributions of all kinds are welcome!  Please submit pull requests, issues, and/or suggestions to the [github repo](https://github.com/RhetTbull/rich_theme_manager).

## Credits

Thank you to [Will McGugan](https://github.com/willmcgugan) for creating [rich](https://github.com/Textualize/rich) and helping to make our command line interfaces more beautiful!

## Projects Using Rich Theme Manager

* [osxphotos](https://github.com/RhetTbull/osxphotos): Python app to export pictures and associated metadata from Apple Photos on macOS. Also includes a package to provide programmatic access to the Photos library, pictures, and metadata.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.11.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.10.0...v0.11.0)

> 1 May 2022

- Bug fix for update [`eb1b363`](https://github.com/RhetTbull/rich_theme_manager/commit/eb1b36311370263b7ab4da648d4d46cf8e396fbb)

#### [v0.10.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.9.0...v0.10.0)

> 1 May 2022

- Bug fix for update [`8499dba`](https://github.com/RhetTbull/rich_theme_manager/commit/8499dba069a513f7f26e94f52a1b85c2ed9f3309)

#### [v0.9.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.8.0...v0.9.0)

> 29 April 2022

- Added update for Theme [`b1761f8`](https://github.com/RhetTbull/rich_theme_manager/commit/b1761f8c519afd9632b8bd2f14464cd969b0884a)

#### [v0.8.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.7.0...v0.8.0)

> 21 March 2022

#### [v0.7.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.6.0...v0.7.0)

> 21 March 2022

- Use Panel for attributes lengend on preview [`19bac69`](https://github.com/RhetTbull/rich_theme_manager/commit/19bac69a28873a7b95da58bffde9cf27e7632e0d)
- Added CONTRIBUTING.md [`3bf5849`](https://github.com/RhetTbull/rich_theme_manager/commit/3bf5849c1cdb6b6d2e7d10d9609f458dcdc005e4)
- version bump [`895a29e`](https://github.com/RhetTbull/rich_theme_manager/commit/895a29e43ddb815a1d52b45b6b0bb866c089242c)

#### [v0.6.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.5.0...v0.6.0)

> 19 March 2022

- Implemented __rich_console_, #5 [`b8e7aa1`](https://github.com/RhetTbull/rich_theme_manager/commit/b8e7aa15b0b005191016d703c499d4cbb0746a7c)
- Added console arg to preview_theme, list_themes, #3 [`a7db52f`](https://github.com/RhetTbull/rich_theme_manager/commit/a7db52f8598f364d2381ecbf347484469935f343)
- Updated example images [`b4ee6dd`](https://github.com/RhetTbull/rich_theme_manager/commit/b4ee6ddb17d0ab0622f7790ea4198c0761ef8246)

#### [v0.5.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.4.0...v0.5.0)

> 16 March 2022

- 100% test coverage [`07ce7a0`](https://github.com/RhetTbull/rich_theme_manager/commit/07ce7a0ae329eac7f8fc3c492abdca164de16ec2)
- Version bump [`b6bb43a`](https://github.com/RhetTbull/rich_theme_manager/commit/b6bb43ac071c5da5572bb08d1877a9d36c640a20)

#### [v0.4.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.3.0...v0.4.0)

> 15 March 2022

- Updated docs, closes #1 [`#1`](https://github.com/RhetTbull/rich_theme_manager/issues/1)
- Updated README, version bump [`1a1f97a`](https://github.com/RhetTbull/rich_theme_manager/commit/1a1f97a8eed01f4b40a943686124a5d903523f12)
- Added .coveragerc [`6992f93`](https://github.com/RhetTbull/rich_theme_manager/commit/6992f931fc72109660a5c2a214e48b6b66a37da8)
- Updated README [`7d5e6d2`](https://github.com/RhetTbull/rich_theme_manager/commit/7d5e6d22d91e20932238100f7af703b7c40f1abe)
- Removed -s [`6f0205a`](https://github.com/RhetTbull/rich_theme_manager/commit/6f0205adba53b0b045cdebb5e88fdfff9f255f3b)
- Added -s to pytest [`e89cb6d`](https://github.com/RhetTbull/rich_theme_manager/commit/e89cb6ddbcb7dadc49db787e6de5d3252a688bd2)

#### [v0.3.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.2.0...v0.3.0)

> 14 March 2022

- Added flake8 [`b66e06c`](https://github.com/RhetTbull/rich_theme_manager/commit/b66e06cd5216e832e1bbcdfe28fdafdc37fde036)
- Create tests [`ecf88a7`](https://github.com/RhetTbull/rich_theme_manager/commit/ecf88a7efc3613f011a664ddd9931ae2e0081d08)
- Added poetry [`4d27661`](https://github.com/RhetTbull/rich_theme_manager/commit/4d276615369f16c34e0fba8d52a2899efced2933)

#### [v0.2.0](https://github.com/RhetTbull/rich_theme_manager/compare/v0.1.2...v0.2.0)

> 14 March 2022

- Added tests, bug fixes [`4326327`](https://github.com/RhetTbull/rich_theme_manager/commit/43263275a4d5c25e8360aafcbfd4ed83d9645f52)
- Initial tests added [`3702502`](https://github.com/RhetTbull/rich_theme_manager/commit/3702502f4f70d0ed47381dba0a16645d9e338450)
- Updated example images [`cdc948e`](https://github.com/RhetTbull/rich_theme_manager/commit/cdc948ee0777c3cb06de68e56dd2325d4745ff09)

#### v0.1.2

> 13 March 2022

- First commit [`1bb50cc`](https://github.com/RhetTbull/rich_theme_manager/commit/1bb50cc24fff6c10701ce69817e67ebc3c554db5)
- Made mypy happy [`0efd838`](https://github.com/RhetTbull/rich_theme_manager/commit/0efd83896d7539ef342a48aa60529d9615b50ce9)
- Updated README [`a91ee96`](https://github.com/RhetTbull/rich_theme_manager/commit/a91ee96047ae44bff74bd6399f1e99821152c91b)


tests/conftest.py
"""test config file for rich_theme_manager"""

import pytest
from rich.style import Style
from rich_theme_manager import Theme

THEMES = [
    Theme(
        name="dark",
        description="Dark mode theme",
        tags=["dark"],
        styles={
            "hidden": Style(color="#383b3d", dim=True),
            "error": Style(color="rgb(255,85,85)", bold=True),
            "filename": Style(color="rgb(189,147,249)", bold=True),
            "filepath": Style(color="rgb(80,250,123)", bold=True),
            "highlight": Style(color="#000000", bgcolor="#d73a49", bold=True),
            "num": Style(color="rgb(139,233,253)", bold=True),
            "time": Style(color="rgb(139,233,253)", bold=True),
            "warning": Style(color="rgb(241,250,140)", bold=True),
        },
    ),
    Theme(
        name="light",
        description="Light mode theme",
        styles={
            "hidden": Style(color="#383b3d", dim=True),
            "error": Style(color="#b31d28", bold=True, underline=True, italic=True),
            "filename": Style(color="#6f42c1", bold=True),
            "filepath": Style(color="#22863a", bold=True),
            "highlight": Style(color="#ffffff", bgcolor="#d73a49", bold=True),
            "num": Style(color="#005cc5", bold=True),
            "time": Style(color="#032f62", bold=True),
            "warning": Style(color="#e36209", bold=True, underline=True, italic=True),
        },
    ),
    Theme(
        name="mono",
        description="Monochromatic theme",
        tags=["mono", "colorblind"],
        styles={
            "hidden": "dim",
            "error": "reverse italic",
            "filename": "bold",
            "filepath": "bold underline",
            "highlight": "reverse italic",
            "num": "bold",
            "time": "bold",
            "warning": "bold italic",
        },
    ),
    Theme(
        name="plain",
        description="Plain theme with no colors",
        tags=["colorblind"],
        styles={
            "hidden": "",
            "error": "",
            "filename": "",
            "filepath": "",
            "highlight": "",
            "num": "",
            "time": "",
            "warning": "",
        },
    ),
]


@pytest.fixture(scope="session")
def themes():
    return THEMES


rich_theme_manager/manager.py
"""Manage rich themes: implements ThemeManager class"""


import pathlib
from os import unlink
from os.path import exists
from typing import Dict, List, Optional

from rich.console import Console
from rich.table import Table

from .theme import SAMPLE_TEXT, Theme


class ThemeManager:
    """Manage rich themes"""

    def __init__(
        self,
        theme_dir: Optional[str] = None,
        themes: Optional[List[Theme]] = None,
        overwrite: bool = False,
        update: bool = False,
    ):
        """Create ThemeManager instance

        Args:
            theme_dir (Optional[str]): directory containing themes
            themes (Optional[List[Theme]]): list of rich_theme_manager.Theme objects
            overwrite (bool): overwrite existing theme files
            update (bool): update existing theme files with new styles from themes but don't replace existing styles
        """
        self._theme_dir: Optional[pathlib.Path] = (
            pathlib.Path(theme_dir) if theme_dir else None
        )
        self._themes: Dict[str, Theme] = (
            {theme.name: theme for theme in themes} if themes else {}
        )

        if self._theme_dir is not None:
            for theme in self.themes:
                theme.path = str(self._theme_dir / f"{theme.name}.theme")

            # if overwrite, need to write themes first to overwrite existing themes
            # then load themes because there may be theme files not in self._themes
            # if not overwrite, themes will be loaded from self._theme_dir then written in case
            # there are themes in self._themes that are not in self._theme_dir
            if overwrite:
                self.write_themes(overwrite=overwrite)
                self.load_themes(update=update)
            else:
                self.load_themes(update=update)
                self.write_themes(overwrite=update)

    @property
    def themes(self) -> List[Theme]:
        """Themes"""
        return list(self._themes.values())

    def add(self, theme: Theme, overwrite=False) -> None:
        """Add theme; if theme file doesn't exist, it will be written"""
        if self._theme_dir and not theme.path:
            theme.path = str(self._theme_dir / f"{theme.name}.theme")
        if self._theme_dir and overwrite or (theme.path and not exists(theme.path)):
            theme.save(overwrite=overwrite)
        self._themes[theme.name] = theme

    def remove(self, theme: Theme) -> None:
        """Remove theme; if theme file exists, it will be deleted"""
        if theme.path and exists(theme.path):
            unlink(theme.path)
        del self._themes[theme.name]

    def get(self, theme_name: str) -> Theme:
        """Get theme by name"""
        for theme in self.themes:
            if theme.name == theme_name:
                return theme
        raise ValueError(f"Theme {theme_name} not found")

    def load_themes(
        self, theme_dir: Optional[str] = None, update: bool = False
    ) -> None:
        """Load themes"""
        theme_dir = theme_dir or self._theme_dir
        if theme_dir is None:
            raise ValueError("No theme directory")

        theme_dir = pathlib.Path(theme_dir)
        for path in theme_dir.glob("*.theme"):
            theme = Theme.read(str(path))
            if update and theme.name in self._themes:
                theme.update(self._themes[theme.name], overwrite_existing_styles=False)
            self._themes[theme.name] = theme

    def write_themes(self, overwrite: bool = False) -> None:
        """Write themes"""
        for theme in self.themes:
            if not theme.path:
                raise ValueError(f"Theme {theme.name} has no path")
            if not exists(theme.path) or overwrite:
                theme.save(overwrite=overwrite)

    def list_themes(
        self,
        show_path: bool = True,
        theme_names: Optional[List[str]] = None,
        console: Optional[Console] = None,
    ) -> None:
        """List themes

        Args:
            show_path (bool): show theme file path
            theme_names (Optional[List[str]]): list of theme names to show (default is all themes)
            console: Optional[Console]: rich.console.Console instance to use for printing
        """
        table = Table(show_header=True, show_lines=False, box=None)
        table.add_column("Theme")
        table.add_column("Description")
        table.add_column("Tags")
        if show_path:
            table.add_column("Path")
        for theme in self.themes:
            if theme_names and theme.name not in theme_names:
                continue
            row = [theme.name, theme.description, ", ".join(theme.tags)]
            if show_path:
                row.append(theme.path or "")
            table.add_row(*row)
        console = console or Console()
        console.print(table)

    @classmethod
    def preview_theme(
        cls,
        theme: Theme,
        sample_text: Optional[str] = None,
        show_path: bool = True,
        console: Optional[Console] = None,
    ) -> None:
        """Preview a theme to the console

        Args:
            theme (Theme): theme to preview
            sample_text (Optional[str]): sample text to use for preview (default is rich_theme_manager.theme.SAMPLE_TEXT)
            show_path (bool): show theme file path
            console: Optional[Console]: rich.console.Console instance to use for printing
        """
        console = console or Console()
        console.print(
            *list(
                theme._preview(
                    sample_text=sample_text or SAMPLE_TEXT, show_path=show_path
                )
            )
        )


rich_theme_manager/__main__.py
"""Example CLI usage of rich_theme_manager.
Uses argparse for no dependencies but should be easily adapatable to click or typer
"""

import argparse
import os
import sys
import tempfile
from textwrap import dedent

from rich.console import Console
from rich.style import Style

# import Theme from rich_theme_manager
from rich_theme_manager import Theme, ThemeManager

# Sample themes for use in the CLI
THEMES = [
    Theme(
        name="dark",
        description="Dark mode theme",
        tags=["dark"],
        styles={
            "hidden": Style(color="#383b3d", dim=True),
            "error": Style(color="rgb(255,85,85)", bold=True),
            "filename": Style(color="rgb(189,147,249)", bold=True),
            "filepath": Style(color="rgb(80,250,123)", bold=True),
            "highlight": Style(color="#000000", bgcolor="#d73a49", bold=True),
            "num": Style(color="rgb(139,233,253)", bold=True),
            "time": Style(color="rgb(139,233,253)", bold=True),
            "warning": Style(color="rgb(241,250,140)", bold=True),
        },
    ),
    Theme(
        name="light",
        description="Light mode theme",
        styles={
            "hidden": Style(color="#383b3d", dim=True),
            "error": Style(color="#b31d28", bold=True, underline=True, italic=True),
            "filename": Style(color="#6f42c1", bold=True),
            "filepath": Style(color="#22863a", bold=True),
            "highlight": Style(color="#ffffff", bgcolor="#d73a49", bold=True),
            "num": Style(color="#005cc5", bold=True),
            "time": Style(color="#032f62", bold=True),
            "warning": Style(color="#e36209", bold=True, underline=True, italic=True),
        },
    ),
    Theme(
        name="mono",
        description="Monochromatic theme",
        tags=["mono", "colorblind"],
        styles={
            "hidden": "dim",
            "error": "reverse italic",
            "filename": "bold",
            "filepath": "bold underline",
            "highlight": "reverse italic",
            "num": "bold",
            "time": "bold",
            "warning": "bold italic",
        },
    ),
    Theme(
        name="plain",
        description="Plain theme with no colors",
        tags=["colorblind"],
        styles={
            "hidden": "",
            "error": "",
            "filename": "",
            "filepath": "",
            "highlight": "",
            "num": "",
            "time": "",
            "warning": "",
        },
    ),
]


def get_theme_dir() -> str:
    """Get the theme directory"""
    # with click, this would be a good place to use get_app_dir
    theme_dir = os.path.join(tempfile.gettempdir(), ".rich_theme_manager")
    if not os.path.exists(theme_dir):
        os.mkdir(theme_dir)
    return theme_dir


def validate_theme_name(name: str) -> None:
    """Check if name is a valid theme name"""
    valid_themes = [theme.name for theme in THEMES]
    if name not in valid_themes:
        print(
            f"'{name}' is not a valid theme name. Valid themes are: {valid_themes}",
            file=sys.stderr,
        )
        sys.exit(1)


def print_example_output(theme: Theme) -> None:
    """Print example output using theme"""
    example_text = f"""
    The following shows examples of how to use custom rich styles with your text.

    This is an example filepath: [filepath]{sys.argv[0]}[/]
    This is an example filename: [filename]{os.path.split(sys.argv[0])[1]}[/]
    This is an example of hidden filename: [hidden].zshrc[/]
    This is an example of a warning: [warning]I've giv'n her all she's got captain, an' I canna give her no more.[/]
    This is an example of an error: [error]I'm sorry, Dave. I'm afraid I can't do that.[/]
    This is an example of a highlight: [highlight]foo[/]
    This is an example of a number: [num]42[/]
    This is an example of a time: [time]12:34[/]
    """
    console = Console(theme=theme, highlight=False)
    console.print(dedent(example_text))


def cli():
    """Example CLI for use of rich_theme_manager"""
    # using argparse so there are no dependencies but dang,
    # does this make me appreciate click and typer better!
    parser = argparse.ArgumentParser(
        prog="rich_theme_manager", description="Example CLI usage of rich_theme_manager"
    )
    parser.add_argument(
        "--example",
        nargs="?",
        const="dark",
        help="Show example output for theme.",
    )
    parser.add_argument(
        "--list",
        help="List themes.",
        action="store_true",
    )
    parser.add_argument(
        "--preview",
        metavar="THEME",
        nargs=1,
        help="Preview theme.",
    )
    parser.add_argument(
        "--config",
        metavar="THEME",
        nargs=1,
        help="Print configuration for theme THEME.",
    )
    args = parser.parse_args()
    theme_manager = ThemeManager(theme_dir=get_theme_dir(), themes=THEMES)

    if not any([args.example, args.list, args.preview, args.config]):
        parser.print_help()
        return

    if args.example:
        validate_theme_name(args.example)
        theme = theme_manager.get(args.example)
        print_example_output(theme)
        return

    if args.list:
        theme_manager.list_themes()
        return

    if args.preview:
        validate_theme_name(args.preview[0])
        theme = theme_manager.get(args.preview[0])
        theme_manager.preview_theme(theme)
        return

    if args.config:
        theme = theme_manager.get(args.config[0])
        print(theme.config)
        return


if __name__ == "__main__":
    cli()
